satpy/writers/awips_tiled.py

```diff
@@ -0,0 +1,1737 @@
+from collections import namedtuple
+
+from satpy.writers import Writer, DecisionTree, Enhancer, get_enhanced_image
+from satpy import __version__
+from pyresample.geometry import AreaDefinition
+from trollsift.parser import StringFormatter, Parser
+
+import numpy as np
+from pyproj import Proj
+import dask
+import dask.array as da
+import xarray as xr
+
+LOG = logging.getLogger(__name__)
+DEFAULT_OUTPUT_PATTERN = '{source_name}_AII_{platform_name}_{sensor}_' \
+                         '{name}_{sector_id}_{tile_id}_' \
+                         '{start_time:%Y%m%d_%H%M}.nc'
+
+UNIT_CONV = {
+    'micron': 'microm',
+    'mm h-1': 'mm/h',
+    '1': '*1',
+    'none': '*1',
+    'percent': '%',
+    'Kelvin': 'kelvin',
+    'K': 'kelvin',
+}
+
+TileInfo = namedtuple('TileInfo', ['tile_count', 'image_shape', 'tile_shape',
+                                   'tile_row_offset', 'tile_column_offset', 'tile_id',
+                                   'tile_number',
+                                   'x', 'y', 'xy_factors', 'tile_slices', 'data_slices'])
+XYFactors = namedtuple('XYFactors', ['mx', 'bx', 'my', 'by'])
+
+
+def fix_awips_file(fn):
+    """Hack the NetCDF4 files to workaround NetCDF-Java bugs used by AWIPS.
+
+    This should not be needed for new versions of AWIPS.
+
+    """
+    # hack to get files created by new NetCDF library
+    # versions to be read by AWIPS buggy java version
+    # of NetCDF
+    LOG.info("Modifying output NetCDF file to work with AWIPS")
+    import h5py
+    h = h5py.File(fn, 'a')
+    if '_NCProperties' in h.attrs:
+        del h.attrs['_NCProperties']
+    h.close()
+
+
+class NumberedTileGenerator(object):
+    """Helper class to generate per-tile metadata for numbered tiles."""
+
+    def __init__(self, area_definition,
+                 tile_shape=None, tile_count=None):
+        """Initialize and generate tile information for this sector/grid for later use."""
+        self.area_definition = area_definition
+        self._rows = self.area_definition.height
+        self._cols = self.area_definition.width
+
+        # get tile shape, number of tiles, etc.
+        self._get_tile_properties(tile_shape, tile_count)
+        # scaling parameters for the overall images X and Y coordinates
+        # they must be the same for all X and Y variables for all tiles
+        # and must be stored in the file as 0, 1, 2, 3, ...
+        # (X factor, X offset, Y factor, Y offset)
+        self.mx, self.bx, self.my, self.by = self._get_xy_scaling_parameters()
+        self.xy_factors = XYFactors(self.mx, self.bx, self.my, self.by)
+        self._tile_cache = []
+
+    def _get_tile_properties(self, tile_shape, tile_count):
+        """Generate tile information for numbered tiles."""
+        if tile_shape is not None:
+            tile_shape = (int(min(tile_shape[0], self._rows)), int(min(tile_shape[1], self._cols)))
+            tile_count = (int(np.ceil(self._rows / float(tile_shape[0]))),
```

Comment:
this seems to be important but isn't tested?

Reply:
Tests added. Good catch.